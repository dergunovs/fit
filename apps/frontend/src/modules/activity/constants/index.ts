import { TActivityChartType } from 'fitness-tracker-contracts';

export const URL_ACTIVITY = '/activities';
export const URL_ACTIVITY_CREATE = `${URL_ACTIVITY}/create`;
export const URL_ACTIVITY_EDIT = `${URL_ACTIVITY}/edit`;

export const URL_ACTIVITY_ADMIN = '/activities_admin';
export const URL_ACTIVITY_ADMIN_EDIT = `${URL_ACTIVITY_ADMIN}/edit`;

export const ACTIVITY_LIST_HEADERS = [{ title: 'Создано' }, { title: 'Пользователь' }];

export const CHART_TYPES: { title: string; value: TActivityChartType }[] = [
  { title: 'Занятия', value: 'activity' },
  { title: 'Подходы', value: 'set' },
  { title: 'Повторы', value: 'repeat' },
  { title: 'Мышцы', value: 'muscle' },
];

export const ACTIVITY_STATISTICS_GAP = 30;
