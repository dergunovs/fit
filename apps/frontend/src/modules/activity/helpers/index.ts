import { IActivity, IExerciseDone, IExerciseStatistics } from 'fitness-tracker-contracts';
import { createTempId, formatDate, formatDuration, subtractDates } from 'mhz-helpers';
import { toast } from 'mhz-ui';

export function getPotentialActivityDuration(
  exercises: IExerciseDone[],
  exerciseStatistics: IExerciseStatistics[],
  averageRestPercent?: number
) {
  if (!averageRestPercent) return '-';

  const totalDuration = exercises.reduce((acc, exercise) => {
    const averageDuration =
      exerciseStatistics.find((choosenExericse) => choosenExericse._id === exercise.exercise?._id)?.averageDuration ||
      0;

    return acc + averageDuration * exercise.repeats;
  }, 0);

  const durationWithRest = Math.round(totalDuration / (1 - averageRestPercent / 100));

  return formatDuration(durationWithRest);
}

export function convertActivityCalendarEvents(activities?: IActivity[]) {
  return activities?.map((activity: IActivity) => {
    return {
      _id: activity._id,
      start: new Date(`${activity.dateCreated}`),
      end: new Date(`${activity.dateUpdated}`),
      title: '+',
      content: activity.exercises,
    };
  });
}

export function generateActivityExercises(exercisesDone: IExerciseDone[]): IExerciseDone[] {
  const activityExercises = exercisesDone.map((exercise) => {
    return {
      _id: createTempId(),
      exercise: {
        _id: exercise.exercise?._id || '',
        title: exercise.exercise?.title || '',
        muscleGroups: exercise.exercise?.muscleGroups || [],
        isWeights: exercise.exercise?.isWeights || false,
        isWeightsRequired: exercise.exercise?.isWeightsRequired || false,
      },
      repeats: exercise.repeats,
      weight: exercise.weight,
    };
  });

  return activityExercises?.length ? [...activityExercises] : [];
}

export function getToFailurePercent(exercises: IExerciseDone[]) {
  const allExercises = exercises.length;
  const toFailureExercises = exercises.filter((exercise) => exercise.isToFailure).length;

  return `${Math.floor((toFailureExercises / allExercises) * 100)}%`;
}

export function getRestPercent(exercises: IExerciseDone[], start?: Date | null | string, end?: Date | null | string) {
  const activityDuration = Number(subtractDates(end, start, true));
  const exercisesDuration = exercises.reduce((acc, current) => acc + (current.duration || 0), 0);

  return `${Math.floor(100 - (exercisesDuration / activityDuration) * 100)}%`;
}

export async function copyActivityToClipboard(
  exercises: IExerciseDone[],
  start?: Date | null | string,
  end?: Date | null | string
) {
  const textHeader = `${formatDate(start, 'ru')}, длительность: ${subtractDates(end, start)}
Подходы: ${exercises.length}, отказы: ${getToFailurePercent(exercises)}, отдых: ${getRestPercent(exercises, start, end)}.

${exercises
  .map((exercise, index) => {
    return `${index + 1}. ${exercise.exercise?.title} x${exercise.repeats} ${exercise.weight ? `${exercise.weight}кг` : ''} ${formatDuration(exercise.duration)} ${exercise.isToFailure ? 'ДО ОТКАЗА' : ''}\n`;
  })
  .join('')}`;

  await navigator.clipboard.writeText(textHeader);

  toast.success('Скопировано в буфер');
}
