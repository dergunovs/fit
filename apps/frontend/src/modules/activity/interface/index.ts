export interface IActivityCalendarEvent<T> {
  _id?: string;
  start: Date | null;
  end: Date | null;
  title: string;
  content: T[];
  color?: string;
}

export interface ICalendarDates {
  dateFrom: string;
  dateTo: string;
}

export interface ITimelineStep {
  left: number;
  right: number;
  type: 'rest' | 'exercise';
}
