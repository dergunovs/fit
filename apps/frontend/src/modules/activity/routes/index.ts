import {
  URL_ACTIVITY_CREATE,
  URL_ACTIVITY_EDIT,
  URL_ACTIVITY_ADMIN,
  URL_ACTIVITY_ADMIN_EDIT,
} from '@/activity/constants';

export const activityRoutes = [
  {
    path: URL_ACTIVITY_CREATE,
    name: 'ActivityCreate',
    component: () => import('@/activity/pages/ActivityCreatePage.vue'),
  },
  {
    path: `${URL_ACTIVITY_EDIT}/:activity`,
    name: 'ActivityPassing',
    component: () => import('@/activity/pages/ActivityPassingPage.vue'),
  },
  {
    path: URL_ACTIVITY_ADMIN,
    name: 'ActivityList',
    component: () => import('@/activity/pages/ActivityListPage.vue'),
    meta: { layout: 'admin' },
  },
  {
    path: `${URL_ACTIVITY_ADMIN_EDIT}/:activity`,
    name: 'ActivityEdit',
    component: () => import('@/activity/pages/ActivityEditPage.vue'),
    meta: { layout: 'admin' },
  },
];
