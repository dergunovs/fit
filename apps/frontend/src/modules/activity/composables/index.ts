import { ref } from 'vue';

import { ICalendarDates } from '@/activity/interface';

export function useActivityCalendar() {
  const dateFrom = ref('');
  const dateTo = ref('');

  const isDatesReady = ref(false);

  function updateDates(dates: ICalendarDates) {
    isDatesReady.value = true;

    dateFrom.value = dates.dateFrom;
    dateTo.value = dates.dateTo;
  }

  return { dateFrom, dateTo, isDatesReady, updateDates };
}
