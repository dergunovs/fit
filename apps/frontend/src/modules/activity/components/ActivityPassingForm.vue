<template>
  <div>
    <UiFlex column>
      <ExercisePassingList
        v-if="props.activity.exercises?.length"
        :exercises="props.activity.exercises"
        :activeExerciseId="activeExerciseId"
        @start="startExercise"
        @stop="stopExercise"
        data-test="exercise-passing-list"
      />

      <div>
        <div>
          Занятие создано <span data-test="activity-start">{{ formatDateTime(props.activity.dateCreated, 'ru') }}</span
          >.
        </div>

        <div v-if="props.activity.dateUpdated">
          Занятие {{ formData.isDone ? 'закончено' : 'обновлено' }}
          <span data-test="activity-updated">{{ formatDateTime(props.activity.dateUpdated, 'ru') }}</span
          >.
        </div>
      </div>

      <UiButton
        @click="finishActivity"
        layout="secondary"
        :isDisabled="props.activity.isDone"
        data-test="activity-finish"
      >
        Завершить занятие досрочно
      </UiButton>
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted, onBeforeUnmount } from 'vue';
import { useRouter } from 'vue-router';
import { toast, UiButton, UiFlex } from 'mhz-ui';
import { formatDateTime, useQueryClient, clone } from 'mhz-helpers';
import {
  API_ACTIVITY,
  API_ACTIVITY_CHART,
  API_ACTIVITY_STATISTICS,
  IActivity,
  IExerciseDone,
} from 'fitness-tracker-contracts';

import ExercisePassingList from '@/exercise/components/ExercisePassingList.vue';

import { activityService } from '@/activity/services';
import { URL_HOME } from '@/common/constants';

interface IProps {
  activity: IActivity;
}

const props = defineProps<IProps>();

const formData = ref<IActivity>({
  exercises: [],
  dateCreated: undefined,
  dateUpdated: undefined,
  dateScheduled: undefined,
  isDone: false,
});

const activeExerciseId = ref<string>();

const router = useRouter();

const queryClient = useQueryClient();

const { mutate: mutateUpdate } = activityService.update({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY] });
  },
});

function startExercise(id: string) {
  activeExerciseId.value = id;
}

function stopExercise(exerciseDone: IExerciseDone) {
  activeExerciseId.value = undefined;

  formData.value.exercises = formData.value.exercises?.map((exercise) => {
    if (exercise._id === exerciseDone._id) {
      return {
        ...exercise,
        isDone: true,
        duration: exerciseDone.duration,
        isToFailure: exerciseDone.isToFailure,
        repeats: exerciseDone.repeats,
        dateUpdated: new Date(),
      };
    }

    return exercise;
  });

  formData.value.isDone = !formData.value.exercises?.some((exercise) => !exercise.isDone);

  if (!formData.value.dateUpdated) {
    formData.value.dateCreated = new Date(new Date().getTime() - (exerciseDone.duration || 0) * 1000);
  }

  formData.value.dateUpdated = new Date();

  mutateUpdate(formData.value);

  if (formData.value.isDone) exitActivity();
}

function finishActivity() {
  formData.value.isDone = true;

  mutateUpdate(formData.value);

  exitActivity();
}

function exitActivity() {
  toast.success('Занятие закончено');

  setTimeout(async () => {
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_STATISTICS] });
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_CHART] });

    router.push(URL_HOME);
  }, 1000);
}

const screenLock = ref<WakeLockSentinel>();

onMounted(async () => {
  if (props.activity) formData.value = clone(props.activity);

  screenLock.value = await navigator.wakeLock?.request();
});

onBeforeUnmount(() => {
  screenLock.value?.release();
  screenLock.value = undefined;
});
</script>
