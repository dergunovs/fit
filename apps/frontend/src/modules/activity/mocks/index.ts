import { ref, computed } from 'vue';
import { vi } from 'vitest';
import { TDeleteActivityDTO, TGetActivitiesDTO } from 'fitness-tracker-contracts';

import * as activityComposables from '@/activity/composables';
import { activityService } from '@/activity/services';
import {
  ACTIVITIES_CALENDAR_FIXTURE,
  ACTIVITIES_FIXTURE,
  ACTIVITIES_STATISTICS_FIXTURE,
  ACTIVITY_FIXTURE,
  ACTIVITIES_CHART_FIXTURE,
} from '@/activity/fixtures';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { IOnSuccess } from '@/common/interface';

const dateFrom = ref('2024-11-24T21:00:00.000Z');
const dateTo = ref('2025-01-05T20:59:59.000Z');
const isDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyGetActivity = vi.spyOn(activityService, 'getOne').mockReturnValue(mockQueryReply(ACTIVITY_FIXTURE));

const getActivitiesData: TGetActivitiesDTO = { data: ACTIVITIES_FIXTURE, total: ACTIVITIES_FIXTURE.length };

const spyGetActivities = vi
  .spyOn(activityService, 'getMany')
  .mockImplementation(() => mockQueryReply(getActivitiesData));

const spyDeleteActivity = vi.fn();

const onSuccess: IOnSuccess = {
  delete: undefined,
};

vi.spyOn(activityService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) onSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteActivityDTO, string>(spyDeleteActivity);
});

const spyUseActivityCalendar = vi.spyOn(activityComposables, 'useActivityCalendar').mockReturnValue({
  dateFrom,
  dateTo,
  isDatesReady,
  updateDates: spyUpdateDates,
});

const computedActivityCalendarEvents = computed(() =>
  activityComposables.convertActivityCalendarEvents(ACTIVITIES_CALENDAR_FIXTURE)
);

const spyComputedActivityCalendarEvents = vi
  .spyOn(activityComposables, 'computedActivityCalendarEvents')
  .mockImplementation(() => computedActivityCalendarEvents);

const spyGetActivitiesCalendar = vi
  .spyOn(activityService, 'getCalendar')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CALENDAR_FIXTURE));

const spyGetActivitiesStatistics = vi
  .spyOn(activityService, 'getStatistics')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_STATISTICS_FIXTURE));

const spyGetActivitiesChart = vi
  .spyOn(activityService, 'getChart')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CHART_FIXTURE));

export {
  spyGetActivities,
  spyGetActivity,
  spyGetActivitiesStatistics,
  spyGetActivitiesChart,
  spyDeleteActivity,
  spyUseActivityCalendar,
  spyComputedActivityCalendarEvents,
  spyGetActivitiesCalendar,
  spyUpdateDates,
  dateFrom,
  dateTo,
  isDatesReady,
  computedActivityCalendarEvents,
  getActivitiesData,
  onSuccess,
};
