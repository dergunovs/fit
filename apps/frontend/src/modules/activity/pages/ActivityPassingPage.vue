<template>
  <div>
    <ActivityPassingForm
      v-if="activity"
      :activity="formData"
      @setDateCreated="(dateCreated) => (formData.dateCreated = dateCreated)"
      @setDateUpdated="(dateUpdated) => (formData.dateUpdated = dateUpdated)"
      @updateExercises="(exercises) => (formData.exercises = exercises)"
      @done="(isDone) => (formData.isDone = isDone)"
      @exit="exitActivity"
      data-test="activity-passing-form"
    />
  </div>
</template>

<script setup lang="ts">
import { ref, watch } from 'vue';
import { useRouter } from 'vue-router';
import { useI18n } from 'vue-i18n';
import { useRouteId, useQueryClient, clone } from 'mhz-helpers';
import { toast } from 'mhz-ui';
import { API_ACTIVITY, API_ACTIVITY_CHART, API_ACTIVITY_STATISTICS, IActivity } from 'fitness-tracker-contracts';

import ActivityPassingForm from '@/activity/components/ActivityPassingForm.vue';

import { activityService } from '@/activity/services';
import { usePageLock } from '@/common/composables';
import { URL_HOME } from '@/common/constants';

const router = useRouter();
const { id } = useRouteId('activity');

const { t } = useI18n();

const queryClient = useQueryClient();

usePageLock();

const { data: activity } = activityService.getOne({}, id);

const formData = ref<IActivity>({
  exercises: [],
  dateCreated: undefined,
  dateUpdated: undefined,
  isDone: false,
});

watch(
  () => activity.value,
  () => {
    if (activity.value) formData.value = clone(activity.value);
  }
);

const { mutate: mutateUpdate } = activityService.update({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY] });
  },
});

function exitActivity() {
  mutateUpdate(formData.value);

  toast.success(t('activity.finished'));

  setTimeout(async () => {
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_STATISTICS] });
    await queryClient.refetchQueries({ queryKey: [API_ACTIVITY_CHART] });

    router.push(URL_HOME);
  }, 1000);
}
</script>
