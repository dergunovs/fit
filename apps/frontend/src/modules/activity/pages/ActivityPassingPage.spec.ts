import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest, wait } from 'mhz-helpers';
import { API_ACTIVITY, API_ACTIVITY_CHART, API_ACTIVITY_STATISTICS } from 'fitness-tracker-contracts';

import ActivityPassingPage from './ActivityPassingPage.vue';
import ActivityPassingForm from '@/activity/components/ActivityPassingForm.vue';

import { wrapperFactory } from '@/common/test';
import { mockOnSuccess, spyGetActivity, spyUpdateActivity } from '@/activity/mocks';
import { mockRouteId, spyRefetchQueries, spyRouterPush, spyToastSuccess, spyUseRouteId } from '@/common/mocks';
import { EXERCISES_DONE_FIXTURE } from '@/exercise/fixtures';
import { URL_HOME } from '@/common/constants';

const form = dataTest('activity-passing-form');

const formData = {
  dateCreated: undefined,
  dateUpdated: undefined,
  exercises: [],
  isDone: false,
};

let wrapper: VueWrapper<InstanceType<typeof ActivityPassingPage>>;

beforeEach(() => {
  wrapper = wrapperFactory(ActivityPassingPage);
});

enableAutoUnmount(afterEach);

describe('ActivityPassingPage', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ActivityPassingPage)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('gets activity', async () => {
    expect(spyUseRouteId).toBeCalledTimes(1);
    expect(spyUseRouteId).toBeCalledWith('activity');

    expect(spyGetActivity).toBeCalledTimes(1);
    expect(spyGetActivity).toBeCalledWith({}, mockRouteId);
  });

  it('updates activity create date', async () => {
    const dateCreated = new Date();

    expect(wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$props.activity).toStrictEqual(formData);

    wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$emit('setDateCreated', dateCreated);

    expect(wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$props.activity).toStrictEqual({
      dateCreated,
      dateUpdated: undefined,
      exercises: [],
      isDone: false,
    });
  });

  it('updates activity update date', async () => {
    const dateUpdated = new Date();

    expect(wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$props.activity).toStrictEqual(formData);

    wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$emit('setDateUpdated', dateUpdated);

    expect(wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$props.activity).toStrictEqual({
      dateCreated: undefined,
      dateUpdated,
      exercises: [],
      isDone: false,
    });
  });

  it('updates activity exercises', async () => {
    expect(wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$props.activity).toStrictEqual(formData);

    wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$emit('updateExercises', EXERCISES_DONE_FIXTURE);

    expect(wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$props.activity).toStrictEqual({
      dateCreated: undefined,
      dateUpdated: undefined,
      exercises: EXERCISES_DONE_FIXTURE,
      isDone: false,
    });
  });

  it('finishes activity', async () => {
    expect(wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$props.activity).toStrictEqual(formData);

    wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$emit('done', true);

    expect(wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$props.activity).toStrictEqual({
      dateCreated: undefined,
      dateUpdated: undefined,
      exercises: [],
      isDone: true,
    });
  });

  it('exits activity page', async () => {
    expect(spyUpdateActivity).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    wrapper.findComponent<typeof ActivityPassingForm>(form).vm.$emit('exit');

    expect(spyUpdateActivity).toBeCalledTimes(1);
    expect(spyUpdateActivity).toBeCalledWith(formData);

    await mockOnSuccess.update?.();

    expect(spyToastSuccess).toBeCalledTimes(1);

    await wait(1000);

    expect(spyRefetchQueries).toBeCalledTimes(3);
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY_STATISTICS] });
    expect(spyRefetchQueries).toBeCalledWith({ queryKey: [API_ACTIVITY_CHART] });

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_HOME);
  });
});
