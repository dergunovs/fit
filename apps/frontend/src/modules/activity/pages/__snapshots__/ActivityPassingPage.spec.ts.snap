// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivityPassingPage > matches snapshot 1`] = `
"<div>
  <activity-passing-form-stub activity="[object Object]" data-test="activity-passing-form"></activity-passing-form-stub>
</div>"
`;
